src/hybridmap/AodScheduler.cpp

```diff
@@ -0,0 +1,549 @@
+    return ActivationMergeType::Append;
+  }
+  return ActivationMergeType::Impossible;
+}
+
+void AodScheduler::AodActivationHelper::reAssignOffsets(
+    std::vector<AodMove*>& aodMoves, int32_t sign) {
+  std::sort(aodMoves.begin(), aodMoves.end(),
+            [](const AodMove* a, const AodMove* b) {
+              return std::abs(a->delta) < std::abs(b->delta);
+            });
+  int32_t offset = sign;
+  for (auto* aodMove : aodMoves) {
+    // same sign
+    if (aodMove->delta * sign >= 0) {
+      aodMove->offset = offset;
+      offset += sign;
+    }
+  }
+}
+
+void AodScheduler::processMoveGroups() {
+  // convert the moves from MoveGroup to AodOperations
+  for (auto groupIt = moveGroups.begin(); groupIt != moveGroups.end();
+       ++groupIt) {
+    AodActivationHelper   aodActivationHelper{arch, OpType::AodActivate};
+    AodActivationHelper   aodDeactivationHelper{arch, OpType::AodDeactivate};
+    MoveGroup             possibleNewMoveGroup{arch};
+    std::vector<AtomMove> movesToRemove;
+    for (auto& movePair : groupIt->moves) {
+      auto& move              = movePair.first;
+      auto  idx               = movePair.second;
+      auto  origin            = arch.getCoordinate(move.first);
+      auto  target            = arch.getCoordinate(move.second);
+      auto  v                 = arch.getVector(move.first, move.second);
+      auto  vReverse          = arch.getVector(move.second, move.first);
+      auto activationCanAddXY = aodActivationHelper.canAddActivation(origin, v);
+      auto deactivationCanAddXY =
+          aodDeactivationHelper.canAddActivation(target, vReverse);
+      if (activationCanAddXY.first == ActivationMergeType::Impossible ||
+          activationCanAddXY.second == ActivationMergeType::Impossible ||
+          deactivationCanAddXY.first == ActivationMergeType::Impossible ||
+          deactivationCanAddXY.second == ActivationMergeType::Impossible) {
+        // move could not be added as not sufficient intermediate levels
+        // add new move group and add move to it
+        possibleNewMoveGroup.add(move, idx);
+        movesToRemove.push_back(move);
+      } else {
+        aodActivationHelper.addActivation(activationCanAddXY, origin, move, v);
+        aodDeactivationHelper.addActivation(deactivationCanAddXY, target, move,
+                                            vReverse);
+      }
+    }
+    // remove from current move group
+    for (const auto& moveToRemove : movesToRemove) {
+      groupIt->moves.erase(
+          std::remove_if(groupIt->moves.begin(), groupIt->moves.end(),
+                         [&moveToRemove](const auto& movePair) {
+                           return movePair.first == moveToRemove;
+                         }),
+          groupIt->moves.end());
+    }
+    if (!possibleNewMoveGroup.moves.empty()) {
+      groupIt = moveGroups.insert(groupIt + 1, std::move(possibleNewMoveGroup));
+      possibleNewMoveGroup = MoveGroup{arch};
+      groupIt--;
+    }
```

Comment:
This code path isn't covered by the existing tests. Is that due to the nature of the tests or is this essentially dead code that could be removed?

Reply:
Is is due to the tests. This happens very occasionally, but we had it during the evaluations of the DAC paper. Unfortunately, I don't remember the settings, and these situations are very hard to create manually.

Reply:
I see. Then, there is no real action item here for now. Would be nice to maybe extend the tests in the future to cover more scenarios. Could just include a couple more circuits 🤷🏼

Reply:
I included only a few to keep the overhead as small as possible. In general one can of course add more.
One could also try around with the random seed, probably there is one which creates some of these special cases.

Reply:
Maybe consider adding one more test circuit and/or architecture for now 🤷🏽‍♂️ it's fine if CI takes a couple seconds longer.